<!DOCTYPE html>
<html lang="en"><head>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-html/tabby.min.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.353">

  <meta name="author" content="Michael Love">
  <title>Tidyomics: Enabling Tidy Data Analysis Workflows for Complex Biological Data</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="site_libs/revealjs/dist/theme/quarto.css">
  <link href="site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <style>
  #title-slide .title {
    font-size: 1.5em
  }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Tidyomics: Enabling Tidy Data Analysis Workflows for Complex Biological Data</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Michael Love 
</div>
        <p class="quarto-title-affiliation">
            Biostatistics &amp; Genetics, UNC-Chapel Hill
          </p>
    </div>
</div>

</section>
<section id="tidyomics-project" class="slide level2">
<h2>Tidyomics project</h2>
<p>An open, open-source project spanning multiple R packages, and developers from around the world. Operates within the Bioconductor Project, with separate funding and organization. Coordination of development via GitHub Projects.</p>
<ul>
<li><p><a href="https://github.com/tidyomics" class="uri">https://github.com/tidyomics</a></p></li>
<li><p><a href="https://www.biorxiv.org/content/10.1101/2023.09.10.557072v2">Announcement paper</a></p></li>
<li><p><code>tidiness_in_bioc</code> channel in Bioconductor Slack</p></li>
</ul>
</section>
<section id="diagram-of-tidyomics-workflows" class="slide level2">
<h2>Diagram of tidyomics workflows</h2>

<img data-src="figures/figure2.png" alt="Diagram of how packages share a similar grammar to operate on data objects. From top to bottom, the data are analyzed, manipulated, and made into plots." class="r-stretch quarto-figure-center"></section>
<section id="international-development-team" class="slide level2">
<h2>International development team</h2>

<img data-src="figures/tidyomics_community.png" alt="Diagram of tidyomics community, with users and developers interacting. On the top are users with arrows coming from developers and packages. On the bottom is the extended community, including Bioconductor." width="300" class="r-stretch quarto-figure-center"></section>
<section id="keep-data-organized-in-objects" class="slide level2">
<h2>Keep data organized in <em>objects</em></h2>
<ul>
<li><p>We typically have more information than just a matrix</p></li>
<li><p>Row and column information (on features and samples)</p></li>
<li><p>Annotated matrix data, i.e.&nbsp;python’s <code>xarray</code>, <code>AnnData</code></p></li>
<li><p>Metadata: organism, genome build, annotation release</p></li>
<li><p>2002: <em>ExpressionSet</em>; 2011: <em>SummarizedExperiment</em></p></li>
<li><p><em>Endomorphic functions</em>: <code>x |&gt; f -&gt; x</code></p></li>
</ul>
</section>
<section id="summarizedexperiment-se" class="slide level2">
<h2>SummarizedExperiment, “SE”</h2>
<p>A <em>SummarizedExperiment</em> is annotated matrix data:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>    s1  s2  s3  s4
g1  94 117 112  94
g2 111 104 103 112
g3  83  87 104 108
g4 101  82 101  99</code></pre>
</div>
</div>
</section>
<section id="row-data" class="slide level2">
<h2>Row data</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 2 columns
           id      symbol
  &lt;character&gt; &lt;character&gt;
1          g1         ABC
2          g2         DEF
3          g3         GHI
4          g4         JKL</code></pre>
</div>
</div>
</section>
<section id="column-data" class="slide level2">
<h2>Column data</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 3 columns
       sample   condition         sex
  &lt;character&gt; &lt;character&gt; &lt;character&gt;
1          s1           x           m
2          s2           y           m
3          s3           x           f
4          s4           y           f</code></pre>
</div>
</div>
</section>
<section id="assembled-object" class="slide level2">
<h2>Assembled object</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 4 4 
metadata(1): organism
assays(1): counts
rownames(4): g1 g2 g3 g4
rowData names(2): id symbol
colnames(4): s1 s2 s3 s4
colData names(3): sample condition sex</code></pre>
</div>
</div>
</section>
<section id="avoids-common-bookkeeping-errors" class="slide level2">
<h2>Avoids common bookkeeping errors</h2>
<p>Reordering samples (columns) is a global operation:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>    s1  s3  s2  s4
g1  94 112 117  94
g2 111 103 104 112
g3  83 104  87 108
g4 101 101  82  99</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 3 columns
        sample   condition         sex
   &lt;character&gt; &lt;character&gt; &lt;character&gt;
s1          s1           x           m
s3          s3           x           f
s2          s2           y           m
s4          s4           y           f</code></pre>
</div>
</div>
<p>Assignment that might result in sample swap results in an error:</p>
<p>Other such validity checks include comparison of genomic ranges across different genome builds: will result in an error.</p>
</section>
<section id="can-be-hard-for-new-users" class="slide level2">
<h2>Can be hard for new users</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "colData"         "assays"          "NAMES"           "elementMetadata" "metadata"       </code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "!=,ANY,Vector-method"                                        
 [2] "!=,Vector,ANY-method"                                        
 [3] "!=,Vector,Vector-method"                                     
 [4] "[,SummarizedExperiment,ANY,ANY,ANY-method"                   
 [5] "[[,SummarizedExperiment,ANY,missing-method"                  
 [6] "[[&lt;-,SummarizedExperiment,ANY,missing-method"                
 [7] "[&lt;-,SummarizedExperiment,ANY,ANY,SummarizedExperiment-method"
 [8] "[&lt;-,Vector,ANY,ANY,ANY-method"                               
 [9] "%in%,ANY,Vector-method"                                      
[10] "%in%,Vector,ANY-method"                                      </code></pre>
</div>
</div>
</section>
<section id="many-beginners-know-dplyrggplot2" class="slide level2">
<h2>Many beginners know <code>dplyr</code>/<code>ggplot2</code></h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  sample condition sex  
  &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;
1 s1     x         m    
2 s3     x         f    </code></pre>
</div>
</div>
</section>
<section id="enabling-dplyr-verbs-for-omics" class="slide level2">
<h2>Enabling dplyr verbs for omics</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble abstraction: 16 × 8
# Features=4 | Samples=4 | Assays=counts
   .feature .sample counts sample condition sex   id    symbol
   &lt;chr&gt;    &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; 
 1 g1       s1          94 s1     x         m     g1    ABC   
 2 g2       s1         111 s1     x         m     g2    DEF   
 3 g3       s1          83 s1     x         m     g3    GHI   
 4 g4       s1         101 s1     x         m     g4    JKL   
 5 g1       s2         117 s2     y         m     g1    ABC   
 6 g2       s2         104 s2     y         m     g2    DEF   
 7 g3       s2          87 s2     y         m     g3    GHI   
 8 g4       s2          82 s2     y         m     g4    JKL   
 9 g1       s3         112 s3     x         f     g1    ABC   
10 g2       s3         103 s3     x         f     g2    DEF   
11 g3       s3         104 s3     x         f     g3    GHI   
12 g4       s3         101 s3     x         f     g4    JKL   
13 g1       s4          94 s4     y         f     g1    ABC   
14 g2       s4         112 s4     y         f     g2    DEF   
15 g3       s4         108 s4     y         f     g3    GHI   
16 g4       s4          99 s4     y         f     g4    JKL   </code></pre>
</div>
</div>
<p>What does this mean “<em>SE-tibble abstraction</em>”?</p>
<p>Essentially this is an API, we can use our familiar verbs and interact with the native object.</p>

<img data-src="figures/counter.png" alt="Picture of a counter with a menu and a bell" class="r-stretch"></section>
<section id="still-a-standard-bioc-object" class="slide level2">
<h2>Still a standard Bioc object</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SummarizedExperiment"
attr(,"package")
[1] "SummarizedExperiment"</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4 4</code></pre>
</div>
</div>
</section>
<section id="we-can-use-familiar-dplyr-verbs" class="slide level2">
<h2>We can use familiar dplyr verbs</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble abstraction: 8 × 8
# Features=4 | Samples=2 | Assays=counts
  .feature .sample counts sample condition sex   id    symbol
  &lt;chr&gt;    &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; 
1 g1       s1          94 s1     x         m     g1    ABC   
2 g2       s1         111 s1     x         m     g2    DEF   
3 g3       s1          83 s1     x         m     g3    GHI   
4 g4       s1         101 s1     x         m     g4    JKL   
5 g1       s3         112 s3     x         f     g1    ABC   
6 g2       s3         103 s3     x         f     g2    DEF   
7 g3       s3         104 s3     x         f     g3    GHI   
8 g4       s3         101 s3     x         f     g4    JKL   </code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 2 rows and 3 columns
        sample   condition         sex
   &lt;character&gt; &lt;character&gt; &lt;character&gt;
s1          s1           x           m
s3          s3           x           f</code></pre>
</div>
</div>
</section>
<section id="facilitates-quick-exploration" class="slide level2">
<h2>Facilitates quick exploration</h2>

<img data-src="tidy-intro-talk_files/figure-revealjs/ggplot-1.png" width="960" class="r-stretch"></section>
<section id="works-with-seurat-and-sce" class="slide level2">
<h2>Works with Seurat and SCE</h2>
<p><em>SingleCellExperiment</em> = <em>SummarizedExperiment</em> with extra bells and whistles for single cells. E.g. slots for reduced dimensions.</p>

<img data-src="figures/umap-1.png" class="r-stretch"></section>
<section id="efficient-operation-on-se-with-plyxp" class="slide level2">
<h2>Efficient operation on SE with <code>plyxp</code></h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SummarizedExperiment"
attr(,"package")
[1] "SummarizedExperiment"</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  a b c  d
A 1 4 7 10
B 2 5 8 11
C 3 6 9 12</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 1 column
      type
  &lt;factor&gt;
a        1
b        1
c        2
d        2</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 3 rows and 1 column
     length
  &lt;numeric&gt;
A        10
B        20
C        30</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble Abstraction: 3 × 4
    .features .samples | counts | length | type 
    &lt;chr&gt;     &lt;chr&gt;    |  &lt;int&gt; |  &lt;dbl&gt; | &lt;fct&gt;
  1 A         a        |      1 |     10 | 1    
  2 B         a        |      2 |     20 | 1    
  3 C         a        |      3 |     30 | 1    
  4 A         b        |      4 |     10 | 1    
  5 B         b        |      5 |     20 | 1    
  … …         …               …        …   …    
n-4 B         c        |      8 |     20 | 2    
n-3 C         c        |      9 |     30 | 2    
n-2 A         d        |     10 |     10 | 2    
n-1 B         d        |     11 |     20 | 2    
n   C         d        |     12 |     30 | 2    
# ℹ n = 12</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble Abstraction: 3 × 4
    .features .samples | counts norm_counts | length | type 
    &lt;chr&gt;     &lt;chr&gt;    |  &lt;int&gt;       &lt;dbl&gt; |  &lt;dbl&gt; | &lt;fct&gt;
  1 A         a        |      1        0.1  |     10 | 1    
  2 B         a        |      2        0.1  |     20 | 1    
  3 C         a        |      3        0.1  |     30 | 1    
  4 A         b        |      4        0.4  |     10 | 1    
  5 B         b        |      5        0.25 |     20 | 1    
  … …         …               …          …         …   …    
n-4 B         c        |      8        0.4  |     20 | 2    
n-3 C         c        |      9        0.3  |     30 | 2    
n-2 A         d        |     10        1    |     10 | 2    
n-1 B         d        |     11        0.55 |     20 | 2    
n   C         d        |     12        0.4  |     30 | 2    
# ℹ n = 12</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble Abstraction: 3 × 2
# Groups: cols(type)
  .features .samples |   sum | length |  type
  &lt;chr&gt;        &lt;int&gt; | &lt;dbl&gt; |  &lt;dbl&gt; | &lt;int&gt;
1 A                1 |     5 |     10 |     1
2 B                1 |     7 |     20 |     1
3 C                1 |     9 |     30 |     1
4 A                2 |    17 |     10 |     2
5 B                2 |    19 |     20 |     2
6 C                2 |    21 |     30 |     2</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  [,1] [,2]
A    5   17
B    7   19
C    9   21</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "PlySummarizedExperiment"
attr(,"package")
[1] "plyxp"</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 1 column
      type
  &lt;factor&gt;
a        1
b        1
c        2
d        2</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 3 4 
metadata(0):
assays(1): counts
rownames(3): A B C
rowData names(1): length
colnames(4): a b c d
colData names(1): type</code></pre>
</div>
</div>
</section>
<section id="computing-on-genomic-ranges-with-tidy-grammar" class="slide level2">
<h2>Computing on genomic ranges with tidy grammar</h2>
<p>Nontrivial to translate biological or biomedical questions about the genome into code. The tidy grammar helps in this case with familiar concepts of filters, joins, grouping, mutation, or summarization.</p>

<img data-src="figures/fillingthegaps.png" alt="Source: 'Filling the Gaps', T2T Consoritium, Science." class="r-stretch"></section>
<section id="genomic-overlap-as-a-join" class="slide level2">
<h2>Genomic overlap as a <code>join</code></h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 40 ranges and 1 metadata column:
       seqnames    ranges strand |     score
          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
   [1]        1   114-115      * |   4.88780
   [2]        1   129-130      * |   4.40817
   [3]        1   154-155      * |   5.18773
   [4]        1   195-196      * |   5.81901
   [5]        1   200-201      * |   4.14720
   ...      ...       ...    ... .       ...
  [36]        1   898-899      * |   6.55006
  [37]        1   922-923      * |   6.46796
  [38]        1   956-957      * |   4.93079
  [39]        1   957-958      * |   4.34292
  [40]        1   966-967      * |   4.53976
  -------
  seqinfo: 1 sequence from an unspecified genome</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 1 metadata column:
      seqnames    ranges strand |          id
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
  [1]        1   101-300      * |           a
  [2]        1   451-650      * |           b
  [3]        1  801-1000      * |           c
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p><img data-src="figures/woodjoin.png" alt="Picture of two stacks of wood being interleaved"></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 30 ranges and 2 metadata columns:
       seqnames    ranges strand |     score          id
          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;character&gt;
   [1]        1   114-115      * |   4.88780           a
   [2]        1   129-130      * |   4.40817           a
   [3]        1   154-155      * |   5.18773           a
   [4]        1   195-196      * |   5.81901           a
   [5]        1   200-201      * |   4.14720           a
   ...      ...       ...    ... .       ...         ...
  [26]        1   898-899      * |   6.55006           c
  [27]        1   922-923      * |   6.46796           c
  [28]        1   956-957      * |   4.93079           c
  [29]        1   957-958      * |   4.34292           c
  [30]        1   966-967      * |   4.53976           c
  -------
  seqinfo: 1 sequence from an unspecified genome</code></pre>
</div>
</div>
<p>Many options, <code>directed</code>, <code>within</code>, <code>maxgap</code>, <code>minoverlap</code>, etc.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 3 rows and 3 columns
           id ave_score         n
  &lt;character&gt; &lt;numeric&gt; &lt;integer&gt;
1           a   5.00465        10
2           b   5.43353        11
3           c   5.45538         7</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="tidy-intro-talk_files/figure-revealjs/ranges-plot-1.png" width="960"></p>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 3 metadata columns:
      seqnames    ranges strand |          id     score  distance
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt; &lt;integer&gt;
  [1]        1   101-102      * |           a   4.88780        11
  [2]        1   451-452      * |           b   3.99047         1
  [3]        1   801-802      * |           c   6.49877        21
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="assess-overlap-significance-using-a-tidy-grammar" class="slide level2">
<h2>Assess overlap significance using a tidy grammar</h2>
<p>something</p>
</section>
<section id="nullranges-for-defining-null-genomic-range-sets" class="slide level2">
<h2>nullranges for defining null genomic range sets</h2>
<p>We developed a package <a href="https://nullranges.github.io/nullranges">nullranges</a>, as a modular tool to assist with making genomic comparisons. It doesn’t do enrichment analysis but provides null genomic range sets for investigating various hypotheses.</p>
</section>
<section id="bootstrapping-ranges" class="slide level2">
<h2>Bootstrapping ranges</h2>
<p>Statistical papers from the ENCODE project noted that <em>block bootstrapping</em> genomic data preserves important spatial patterns (Bickel <em>et al.</em> 2010).</p>
<p><img data-src="figures/boot.png" alt="Diagram of block bootstrapping genomic ranges. Blocks are resampled from original data and arranged to form new range sets."></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 60 rows and 3 columns
     iter          id n_overlaps
    &lt;Rle&gt; &lt;character&gt;  &lt;integer&gt;
1       1           a          4
2       1           b         10
3       1           c          4
4       2           a         11
5       2           b         10
...   ...         ...        ...
56     19           b         18
57     19           c         11
58     20           a         10
59     20           b         15
60     20           c          8</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="tidy-intro-talk_files/figure-revealjs/boot-plot-1.png" width="960"></p>
</div>
</div>
</section>
<section id="matching-ranges" class="slide level2">
<h2>Matching ranges</h2>
<p>Matching on covariates from a large pool allows for more focused hypothesis testing.</p>
<p><img data-src="figures/match.png" alt="Diagram of matching genomic ranges. A pool of different colored ranges are drawn from to match the warmer colors of a focal set of ranges."></p>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="tidy-intro-talk_files/figure-revealjs/match-covar-plot-1.png" width="960"></p>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 82 ranges and 2 metadata columns:
       seqnames    ranges strand |     score origin
          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;  &lt;Rle&gt;
   [1]        1   154-155      * |   5.91030  focal
   [2]        1   195-196      * |   6.95693  focal
   [3]        1   280-281      * |   6.62346  focal
   [4]        1   282-283      * |   5.20914  focal
   [5]        1   303-304      * |   7.47406  focal
   ...      ...       ...    ... .       ...    ...
  [78]        1   898-899      * |   6.55006   pool
  [79]        1   922-923      * |   6.46796   pool
  [80]        1   956-957      * |   4.93079   pool
  [81]        1   957-958      * |   4.34292   pool
  [82]        1   966-967      * |   4.53976   pool
  -------
  seqinfo: 1 sequence from an unspecified genome</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 9 rows and 3 columns
           id   origin ave_score
  &lt;character&gt;    &lt;Rle&gt; &lt;numeric&gt;
1           a    focal   6.17495
2           a matching   5.54412
3           a     pool   5.00465
4           b    focal   5.79665
5           b matching   6.07625
6           b     pool   5.25055
7           c    focal   6.32536
8           c matching   5.82463
9           c     pool   5.14840</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="tidy-intro-talk_files/figure-revealjs/match-overlap-plot-1.png" width="960"></p>
</div>
</div>
<p>These are published as Application Notes:</p>
<ul>
<li><p><a href="https://doi.org/10.1093/bioinformatics/btad190">bootRanges</a></p></li>
<li><p><a href="https://doi.org/10.1093/bioinformatics/btad197">matchRanges</a></p></li>
</ul>
</section>
<section id="current-state-and-future-directions" class="slide level2">
<h2>Current state and future directions</h2>
<p>something</p>
</section>
<section id="what-has-been-implemented" class="slide level2">
<h2>What has been implemented</h2>
<ul>
<li><p>Matrix-shaped objects (SE, SCE)</p></li>
<li><p>Ranges</p></li>
<li><p>Interactions</p></li>
<li><p>Cytometry</p></li>
<li><p>Spatial</p></li>
<li><p>more to come…</p></li>
</ul>
</section>
<section id="limitations" class="slide level2">
<h2>Limitations</h2>
<ul>
<li>package code and <em>non-standard evaluation</em></li>
<li>optimized code, e.g.&nbsp;matrix operations</li>
</ul>
</section>
<section id="how-to-contribute" class="slide level2">
<h2>How to contribute</h2>
<ul>
<li><p><a href="https://github.com/tidyomics" class="uri">https://github.com/tidyomics</a>, read docs, see open challenges</p></li>
<li><p>Motivation: <a href="https://www.biorxiv.org/content/10.1101/2023.09.10.557072v2" class="uri">https://www.biorxiv.org/content/10.1101/2023.09.10.557072v2</a></p></li>
<li><p><code>tidiness_in_bioc</code> channel in Bioconductor Slack</p></li>
</ul>
</section>
<section id="outro" class="slide level2">
<h2>Outro</h2>
<p>Omics data analysis means looking at:</p>
<ul>
<li>main contributions to variance (e.g.&nbsp;PCA, see <code>plotPCA</code> for bulk and <a href="https://bioconductor.org/books/release/OSCA/">OSCA</a> for sc), also <code>variancePartition</code></li>
<li>column and row densities (<code>tidySE</code> allows directly plotting <code>geom_density</code> of rows/columns, or <code>geom_violin</code>)</li>
<li>known positive features, feature-level plots (<code>filter</code> to feature, pipe to <code>geom_point</code> etc.)</li>
</ul>
<p>If you’re interested in more complicated use cases of <code>tidyomics</code> see this online book:</p>
<ul>
<li><a href="https://tidyomics.github.io/tidy-ranges-tutorial">Tidy ranges tutorial</a></li>
</ul>
</section>
<section id="contributors" class="slide level2">
<h2>Contributors</h2>
<ul>
<li><p>Stefano Mangiola (<em>tidyomics</em> leadership, tidy expression, single cell, spatial)</p></li>
<li><p>Justin Landis (<em>plyxp</em>)</p></li>
<li><p>Eric Davis, Wancen Mu, Doug Phanstiel (<em>nullranges</em>)</p></li>
<li><p>Stuart Lee, Michael Lawrence, Di Cook (<em>plyranges</em>)</p></li>
</ul>
<p>And <strong>tidyomics developers</strong>: William Hutchison, Timothy Keyes, Helena Crowell, Jacques Serizay, Charlotte Soneson, Eric Davis, Noriaki Sato, Lambda Moses, Boyd Tarlinton, Abdullah Nahid, Miha Kosmac, Quentin Clayssen, Victor Yuan, Wancen Mu, Ji-Eun Park, Izabela Mamede, Min Hyung Ryu, Pierre-Paul Axisa, Paulina Paiz, Chi-Lam Poon, Ming Tang</p>
</section>
<section id="funding" class="slide level2">
<h2>Funding</h2>
<p><em>tidyomics</em> project funded by an Essential Open Source Software award from CZI and Wellcome Trust</p>

<div class="footer footer-default">

</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="site_libs/revealjs/plugin/search/search.js"></script>
  <script src="site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>