<!DOCTYPE html>
<html lang="en"><head>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-html/tabby.min.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.40">

  <meta name="author" content="Michael Love">
  <title>Tidyomics: Enabling Tidy Data Analysis for Complex Biological Data</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="site_libs/revealjs/dist/theme/quarto-423d414d36ce366a9d05f36d06304417.css">
  <link href="site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <style>
  #title-slide .title {
    font-size: 1.5em
  }
  .cell-output {
    margin-top: 20px;
  }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Tidyomics: Enabling Tidy Data Analysis for Complex Biological Data</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Michael Love 
</div>
        <p class="quarto-title-affiliation">
            Dept of Genetics, Dept of Biostatistics, UNC-Chapel Hill
          </p>
    </div>
</div>

</section>
<section id="tidyomics-project" class="slide level2">
<h2>Tidyomics project</h2>
<p>An open, open-source project spanning multiple R packages, and developers from around the world. Operates within the Bioconductor Project, with separate funding and organization. Coordination of development via GitHub Projects.</p>
<ul>
<li><p><a href="https://github.com/tidyomics" class="uri">https://github.com/tidyomics</a></p></li>
<li><p><a href="https://www.biorxiv.org/content/10.1101/2023.09.10.557072v2">Announcement paper</a></p></li>
<li><p><code>#tidiness_in_bioc</code> channel in Zulip</p></li>
</ul>
</section>
<section id="what-is-tidy-data" class="slide level2">
<h2>What is “tidy data”</h2>
<ul>
<li><p>One row per observation, one column per variable.</p></li>
<li><p>Genomic regions (BED) already in this format.</p></li>
<li><p>Matrices and annotated matrices are not.</p></li>
</ul>
</section>
<section id="a-grammar-of-data-manipulation" class="slide level2">
<h2>A grammar of data manipulation</h2>
<style>
  .small-text ul {
    font-size: 0.8em;
  }
</style>
<p>In the R package <em>dplyr</em>:</p>
<div class="small-text">
<ul>
<li><code>mutate()</code> adds new variables that are functions of existing variables.</li>
<li><code>select()</code> picks variables based on their names.</li>
<li><code>filter()</code> picks cases based on their values.</li>
<li><code>slice()</code> picks cases based on their position.</li>
<li><code>summarize()</code> reduces multiple values down to a single summary.</li>
<li><code>arrange()</code> changes the ordering of the rows.</li>
<li><code>group_by()</code> perform any operation by group.</li>
</ul>
</div>
<p><a href="https://dplyr.tidyverse.org/" class="uri">https://dplyr.tidyverse.org/</a></p>
</section>
<section id="the-pipe" class="slide level2">
<h2>The pipe</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a></a><span class="bu">command</span> <span class="kw">|</span> <span class="bu">command</span> <span class="kw">|</span> <span class="bu">command</span> <span class="op">&gt;</span> output.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<blockquote>
<p>“Pipes rank alongside the hierarchical file system and regular expressions as one of the most powerful yet elegant features of Unix-like operating systems.”</p>
</blockquote>
<p><a href="http://www.linfo.org/pipe.html" class="uri">http://www.linfo.org/pipe.html</a></p>
<p>In R we use <code>%&gt;%</code> or <code>|&gt;</code> instead of <code>|</code> to chain operations.</p>
</section>
<section id="tidyomics-dplyr-verbs-applied-to-omics" class="slide level2">
<h2>Tidyomics = <em>dplyr</em> verbs applied to omics</h2>
<ul>
<li><p>Genomic regions (called “ranges”)</p></li>
<li><p>Matrix data, e.g.&nbsp;gene expression over samples or cells</p></li>
<li><p>Genomic interactions, e.g.&nbsp;DNA loops</p></li>
<li><p>And more…</p></li>
</ul>
</section>
<section id="diagram-of-tidyomics-workflows" class="slide level2">
<h2>Diagram of tidyomics workflows</h2>

<img data-src="figures/figure2.png" class="quarto-figure quarto-figure-center r-stretch" alt="Diagram of how packages share a similar grammar="></section>
<section id="international-development-team" class="slide level2">
<h2>International development team</h2>

<img data-src="figures/tidyomics_community.png" class="quarto-figure quarto-figure-center r-stretch" alt="Diagram of tidyomics community=" width="300"></section>
<section id="keep-data-organized-in-objects" class="slide level2">
<h2>Keep data organized in <em>objects</em></h2>
<ul>
<li><p>We typically have more information than just a matrix</p></li>
<li><p>Row and column information (on features and samples)</p></li>
<li><p>Annotated matrix data, i.e.&nbsp;python’s <code>xarray</code>, <code>AnnData</code></p></li>
<li><p>Metadata: organism, genome build, annotation release</p></li>
<li><p>2002: <em>ExpressionSet</em>; 2011: <em>SummarizedExperiment</em></p></li>
<li><p><em>Endomorphic functions</em>: <code>x |&gt; f -&gt; x</code></p></li>
</ul>
</section>
<section id="summarizedexperiment-se" class="slide level2">
<h2>SummarizedExperiment, “SE”</h2>
<p>A <em>SummarizedExperiment</em> is annotated matrix data.</p>
<p>Imagine a matrix of counts:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a>counts <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb2-2"><a></a>  <span class="fu">rpois</span>(<span class="dv">16</span>, <span class="at">lambda=</span><span class="dv">100</span>), <span class="at">ncol=</span><span class="dv">4</span>,</span>
<span id="cb2-3"><a></a>  <span class="at">dimnames=</span><span class="fu">list</span>(<span class="fu">c</span>(<span class="st">"g1"</span>,<span class="st">"g2"</span>,<span class="st">"g3"</span>,<span class="st">"g4"</span>),</span>
<span id="cb2-4"><a></a>                <span class="fu">c</span>(<span class="st">"s1"</span>,<span class="st">"s2"</span>,<span class="st">"s3"</span>,<span class="st">"s4"</span>))</span>
<span id="cb2-5"><a></a>)</span>
<span id="cb2-6"><a></a>counts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    s1  s2  s3  s4
g1  94 117 112  94
g2 111 104 103 112
g3  83  87 104 108
g4 101  82 101  99</code></pre>
</div>
</div>
</section>
<section id="row-data-and-column-data" class="slide level2">
<h2>Row data and column data</h2>
<p>metadata about genes (rows)</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 2 columns
           id      symbol
  &lt;character&gt; &lt;character&gt;
1          g1         ABC
2          g2         DEF
3          g3         GHI
4          g4         JKL</code></pre>
</div>
</div>
<p>metadata about samples (columns)</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 3 columns
       sample   condition         sex
  &lt;character&gt; &lt;character&gt; &lt;character&gt;
1          s1           x           m
2          s2           y           m
3          s3           x           f
4          s4           y           f</code></pre>
</div>
</div>
</section>
<section id="assembled-object" class="slide level2">
<h2>Assembled object</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(</span>
<span id="cb6-2"><a></a>  <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> counts),</span>
<span id="cb6-3"><a></a>  <span class="at">rowData =</span> genes, </span>
<span id="cb6-4"><a></a>  <span class="at">colData =</span> samples,</span>
<span id="cb6-5"><a></a>  <span class="at">metadata =</span> <span class="fu">list</span>(<span class="at">organism=</span><span class="st">"Hsapiens"</span>)</span>
<span id="cb6-6"><a></a>)</span>
<span id="cb6-7"><a></a>se</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 4 4 
metadata(1): organism
assays(1): counts
rownames(4): g1 g2 g3 g4
rowData names(2): id symbol
colnames(4): s1 s2 s3 s4
colData names(3): sample condition sex</code></pre>
</div>
</div>
</section>
<section id="deals-with-bookkeeping-issues" class="slide level2">
<h2>Deals with bookkeeping issues</h2>
<p>Reordering columns propagates to <code>assay</code> and <code>colData</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a></a>se2 <span class="ot">&lt;-</span> se[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">4</span>)]</span>
<span id="cb8-2"><a></a><span class="fu">assay</span>(se2, <span class="st">"counts"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    s1  s3  s2  s4
g1  94 112 117  94
g2 111 103 104 112
g3  83 104  87 108
g4 101 101  82  99</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a></a><span class="fu">colData</span>(se2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 3 columns
        sample   condition         sex
   &lt;character&gt; &lt;character&gt; &lt;character&gt;
s1          s1           x           m
s3          s3           x           f
s2          s2           y           m
s4          s4           y           f</code></pre>
</div>
</div>
</section>
<section id="deals-with-bookkeeping-issues-1" class="slide level2">
<h2>Deals with bookkeeping issues</h2>
<p>Assignment with conflicting metadata gives an error:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a></a><span class="fu">assay</span>(se2) <span class="ot">&lt;-</span> counts</span>
<span id="cb12-2"><a></a><span class="co"># Error:</span></span>
<span id="cb12-3"><a></a><span class="co"># please use 'assay(x, withDimnames=FALSE)) &lt;- value' or </span></span>
<span id="cb12-4"><a></a><span class="co"># 'assays(x, withDimnames=FALSE)) &lt;- value'</span></span>
<span id="cb12-5"><a></a><span class="co"># when the rownames or colnames of the supplied assay(s) are not </span></span>
<span id="cb12-6"><a></a><span class="co"># identical to those of the receiving  SummarizedExperiment object 'x'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p>Other such validity checks include comparison across <em>different genome builds</em>.</p></li>
<li><p>Errors triggered by metadata better than silent errors!</p></li>
</ul>
</section>
<section id="can-be-hard-for-new-users" class="slide level2">
<h2>Can be hard for new users</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a></a><span class="fu">slotNames</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "colData"         "assays"          "NAMES"           "elementMetadata" "metadata"       </code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a></a><span class="fu">methods</span>(<span class="at">class =</span> <span class="fu">class</span>(se)) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "!=,ANY,Vector-method"                         "!=,Vector,ANY-method"                        
[3] "!=,Vector,Vector-method"                      "[,SummarizedExperiment,ANY,ANY,ANY-method"   
[5] "[[,SummarizedExperiment,ANY,missing-method"   "[[&lt;-,SummarizedExperiment,ANY,missing-method"</code></pre>
</div>
</div>
<ul>
<li><p>An advanced R/Bioconductor user should eventually learn these methods, class/method inheritance.</p></li>
<li><p>Not needed to explore or visualize data, or make basic data summaries.</p></li>
</ul>
</section>
<section id="beginners-know-dplyr-ggplot2" class="slide level2">
<h2>Beginners know <em>dplyr</em> &amp; <em>ggplot2</em></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a></a><span class="fu">library</span>(dplyr)</span>
<span id="cb17-2"><a></a><span class="co"># filter to samples in condition 'x'</span></span>
<span id="cb17-3"><a></a>samples <span class="sc">|&gt;</span> </span>
<span id="cb17-4"><a></a>  <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span> </span>
<span id="cb17-5"><a></a>  <span class="fu">filter</span>(condition <span class="sc">==</span> <span class="st">"x"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  sample condition sex  
  &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;
1 s1     x         m    
2 s3     x         f    </code></pre>
</div>
</div>
</section>
<section id="enabling-dplyr-verbs-for-omics" class="slide level2">
<h2>Enabling <em>dplyr</em> verbs for omics</h2>
<ul>
<li><p><em>tidySummarizedExperiment</em> package from Mangiola <em>et al.</em></p></li>
<li><p>Printing says: “<em>SummarizedExperiment-tibble abstraction</em>”</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a></a><span class="fu">library</span>(tidySummarizedExperiment)</span>
<span id="cb19-2"><a></a>se</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble abstraction: 16 × 8
# Features=4 | Samples=4 | Assays=counts
   .feature .sample counts sample condition sex   id    symbol
   &lt;chr&gt;    &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; 
 1 g1       s1          94 s1     x         m     g1    ABC   
 2 g2       s1         111 s1     x         m     g2    DEF   
 3 g3       s1          83 s1     x         m     g3    GHI   
 4 g4       s1         101 s1     x         m     g4    JKL   
 5 g1       s2         117 s2     y         m     g1    ABC   
 6 g2       s2         104 s2     y         m     g2    DEF   
 7 g3       s2          87 s2     y         m     g3    GHI   
 8 g4       s2          82 s2     y         m     g4    JKL   
 9 g1       s3         112 s3     x         f     g1    ABC   
10 g2       s3         103 s3     x         f     g2    DEF   
11 g3       s3         104 s3     x         f     g3    GHI   
12 g4       s3         101 s3     x         f     g4    JKL   
13 g1       s4          94 s4     y         f     g1    ABC   
14 g2       s4         112 s4     y         f     g2    DEF   
15 g3       s4         108 s4     y         f     g3    GHI   
16 g4       s4          99 s4     y         f     g4    JKL   </code></pre>
</div>
</div>
</section>
<section id="tidyomics-is-an-api" class="slide level2">
<h2>Tidyomics is an API</h2>
<p>Interact with native objects using standard R/Bioc methods.</p>

<img data-src="figures/counter.png" class="quarto-figure quarto-figure-center r-stretch" alt="Counter with a menu and a bell"></section>
<section id="still-a-standard-bioc-object" class="slide level2">
<h2>Still a standard Bioc object</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a></a><span class="fu">class</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "SummarizedExperiment"
attr(,"package")
[1] "SummarizedExperiment"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a></a><span class="fu">dim</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4 4</code></pre>
</div>
</div>
</section>
<section id="we-can-use-familiar-dplyr-verbs" class="slide level2">
<h2>We can use familiar dplyr verbs</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a></a>se <span class="sc">|&gt;</span> </span>
<span id="cb25-2"><a></a>  <span class="fu">filter</span>(condition <span class="sc">==</span> <span class="st">"x"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble abstraction: 8 × 8
# Features=4 | Samples=2 | Assays=counts
  .feature .sample counts sample condition sex   id    symbol
  &lt;chr&gt;    &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; 
1 g1       s1          94 s1     x         m     g1    ABC   
2 g2       s1         111 s1     x         m     g2    DEF   
3 g3       s1          83 s1     x         m     g3    GHI   
4 g4       s1         101 s1     x         m     g4    JKL   
5 g1       s3         112 s3     x         f     g1    ABC   
6 g2       s3         103 s3     x         f     g2    DEF   
7 g3       s3         104 s3     x         f     g3    GHI   
8 g4       s3         101 s3     x         f     g4    JKL   </code></pre>
</div>
</div>
</section>
<section id="we-can-use-familiar-dplyr-verbs-1" class="slide level2">
<h2>We can use familiar dplyr verbs</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a></a>se <span class="sc">|&gt;</span> </span>
<span id="cb27-2"><a></a>  <span class="fu">filter</span>(condition <span class="sc">==</span> <span class="st">"x"</span>) <span class="sc">|&gt;</span></span>
<span id="cb27-3"><a></a>  <span class="fu">colData</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 2 rows and 3 columns
        sample   condition         sex
   &lt;character&gt; &lt;character&gt; &lt;character&gt;
s1          s1           x           m
s3          s3           x           f</code></pre>
</div>
</div>
</section>
<section id="facilitates-quick-exploration" class="slide level2">
<h2>Facilitates quick exploration</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb29-2"><a></a>se <span class="sc">|&gt;</span> </span>
<span id="cb29-3"><a></a>  <span class="fu">filter</span>(.feature <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"g1"</span>,<span class="st">"g2"</span>)) <span class="sc">|&gt;</span> <span class="co"># `.feature` a special name</span></span>
<span id="cb29-4"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(condition, counts, <span class="at">color=</span>sex, <span class="at">group=</span>sex)) <span class="sc">+</span> </span>
<span id="cb29-5"><a></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_line</span>() <span class="sc">+</span> <span class="fu">facet_wrap</span>(<span class="sc">~</span>.feature)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="tidy-intro-talk_files/figure-revealjs/ggplot-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="seurat-and-singlecellexperiment" class="slide level2">
<h2>Seurat and SingleCellExperiment</h2>
<p><em>SingleCellExperiment</em> has slots for e.g.&nbsp;reduced dimensions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a></a>sce <span class="sc">|&gt;</span></span>
<span id="cb30-2"><a></a>  <span class="fu">filter</span>(Phase <span class="sc">==</span> <span class="st">"G1"</span>) <span class="sc">|&gt;</span></span>
<span id="cb30-3"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(UMAP_1, UMAP_2, <span class="at">color=</span>nCount_RNA)) <span class="sc">+</span></span>
<span id="cb30-4"><a></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">scale_color_gradient2</span>(<span class="at">trans=</span><span class="st">"log10"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<img data-src="figures/umap-1.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="efficient-operation-on-se-with-plyxp" class="slide level2">
<h2>Efficient operation on SE with <code>plyxp</code></h2>
<ul>
<li><p>Justin Landis (UNC BCB) noticed some opportunities for more efficient operations.</p></li>
<li><p>Also, reduce ambiguity, allow multiple ways to access data across context. Leveraging data masks from <em>rlang</em>.</p></li>
<li><p>Developed in Summer 2024: <code>plyxp</code>, stand-alone but also as a <em>tidySummarizedExperiment</em> engine.</p></li>
</ul>

<img data-src="figures/plyxp.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="efficient-operation-on-se-with-plyxp-1" class="slide level2">
<h2>Efficient operation on SE with <code>plyxp</code></h2>

<img data-src="figures/plyxp-bindings.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="suppose-the-following-se-object" class="slide level2">
<h2>Suppose the following SE object</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a></a><span class="fu">assay</span>(simple)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  a b c  d
A 1 4 7 10
B 2 5 8 11
C 3 6 9 12</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a></a><span class="fu">colData</span>(simple) <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  type
a    1
b    1
c    2
d    2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a></a><span class="fu">rowData</span>(simple) <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  length
A     10
B     20
C     30</code></pre>
</div>
</div>
</section>
<section id="trigger-methods-with-new_plyxp" class="slide level2">
<h2>Trigger methods with <code>new_plyxp()</code></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a></a><span class="fu">library</span>(plyxp)</span>
<span id="cb37-2"><a></a>xp <span class="ot">&lt;-</span> simple <span class="sc">|&gt;</span> </span>
<span id="cb37-3"><a></a>  <span class="fu">new_plyxp</span>()</span>
<span id="cb37-4"><a></a>xp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble Abstraction: 3 × 4
    .features .samples | counts | length | type 
    &lt;chr&gt;     &lt;chr&gt;    |  &lt;int&gt; |  &lt;dbl&gt; | &lt;fct&gt;
  1 A         a        |      1 |     10 | 1    
  2 B         a        |      2 |     20 | 1    
  3 C         a        |      3 |     30 | 1    
  4 A         b        |      4 |     10 | 1    
  5 B         b        |      5 |     20 | 1    
  … …         …               …        …   …    
n-4 B         c        |      8 |     20 | 2    
n-3 C         c        |      9 |     30 | 2    
n-2 A         d        |     10 |     10 | 2    
n-1 B         d        |     11 |     20 | 2    
n   C         d        |     12 |     30 | 2    
# ℹ n = 12</code></pre>
</div>
</div>
</section>
<section id="mutate-with-plyxp" class="slide level2">
<h2>Mutate with <code>plyxp</code></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a></a>xp <span class="sc">|&gt;</span> <span class="co"># assay is default context for mutate()</span></span>
<span id="cb39-2"><a></a>  <span class="fu">mutate</span>(<span class="at">norm_counts =</span> counts <span class="sc">/</span> .rows<span class="sc">$</span>length)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble Abstraction: 3 × 4
    .features .samples | counts norm_counts | length | type 
    &lt;chr&gt;     &lt;chr&gt;    |  &lt;int&gt;       &lt;dbl&gt; |  &lt;dbl&gt; | &lt;fct&gt;
  1 A         a        |      1        0.1  |     10 | 1    
  2 B         a        |      2        0.1  |     20 | 1    
  3 C         a        |      3        0.1  |     30 | 1    
  4 A         b        |      4        0.4  |     10 | 1    
  5 B         b        |      5        0.25 |     20 | 1    
  … …         …               …          …         …   …    
n-4 B         c        |      8        0.4  |     20 | 2    
n-3 C         c        |      9        0.3  |     30 | 2    
n-2 A         d        |     10        1    |     10 | 2    
n-1 B         d        |     11        0.55 |     20 | 2    
n   C         d        |     12        0.4  |     30 | 2    
# ℹ n = 12</code></pre>
</div>
</div>
</section>
<section id="what-was-that-.rows-doing" class="slide level2">
<h2>What was that <code>.rows$</code> doing?</h2>

<img data-src="figures/plxyp-pronouns.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="grammar-for-group-and-summarize" class="slide level2">
<h2>Grammar for group and summarize</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a></a>summed <span class="ot">&lt;-</span> xp <span class="sc">|&gt;</span></span>
<span id="cb41-2"><a></a>  <span class="fu">group_by</span>(<span class="fu">cols</span>(type)) <span class="sc">|&gt;</span></span>
<span id="cb41-3"><a></a>  <span class="fu">summarize</span>(<span class="at">sum =</span> <span class="fu">rowSums</span>(counts))</span>
<span id="cb41-4"><a></a>summed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble Abstraction: 3 × 2
# Groups: cols(type)
  .features .samples |   sum | length |  type
  &lt;chr&gt;        &lt;int&gt; | &lt;dbl&gt; |  &lt;dbl&gt; | &lt;int&gt;
1 A                1 |     5 |     10 |     1
2 B                1 |     7 |     20 |     1
3 C                1 |     9 |     30 |     1
4 A                2 |    17 |     10 |     2
5 B                2 |    19 |     20 |     2
6 C                2 |    21 |     30 |     2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a></a><span class="fu">assay</span>(summed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [,1] [,2]
A    5   17
B    7   19
C    9   21</code></pre>
</div>
</div>
</section>
<section id="multiple-implementations" class="slide level2">
<h2>Multiple implementations</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a></a>xp <span class="sc">|&gt;</span> <span class="co"># .assays_asis gives direct access to the matrix</span></span>
<span id="cb45-2"><a></a>  <span class="fu">mutate</span>(<span class="fu">rows</span>(<span class="at">ave_counts =</span> <span class="fu">rowMeans</span>(.assays_asis<span class="sc">$</span>counts)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A SummarizedExperiment-tibble Abstraction: 3 × 4
    .features .samples | counts | length ave_counts | type 
    &lt;chr&gt;     &lt;chr&gt;    |  &lt;int&gt; |  &lt;dbl&gt;      &lt;dbl&gt; | &lt;fct&gt;
  1 A         a        |      1 |     10        5.5 | 1    
  2 B         a        |      2 |     20        6.5 | 1    
  3 C         a        |      3 |     30        7.5 | 1    
  4 A         b        |      4 |     10        5.5 | 1    
  5 B         b        |      5 |     20        6.5 | 1    
  … …         …               …        …         …    …    
n-4 B         c        |      8 |     20        6.5 | 2    
n-3 C         c        |      9 |     30        7.5 | 2    
n-2 A         d        |     10 |     10        5.5 | 2    
n-1 B         d        |     11 |     20        6.5 | 2    
n   C         d        |     12 |     30        7.5 | 2    
# ℹ n = 12</code></pre>
</div>
</div>
<ul>
<li>See vignette for more: <a href="https://jtlandis.github.io/plyxp/" class="uri">https://jtlandis.github.io/plyxp/</a></li>
</ul>
</section>
<section id="the-plyxp-class" class="slide level2">
<h2>The <code>plyxp</code> class</h2>
<p>A simple wrapper of SE:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a></a><span class="fu">class</span>(xp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "PlySummarizedExperiment"
attr(,"package")
[1] "plyxp"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a></a><span class="fu">colData</span>(xp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 1 column
      type
  &lt;factor&gt;
a        1
b        1
c        2
d        2</code></pre>
</div>
</div>
</section>
<section id="unwrapping-plyxp" class="slide level2">
<h2>Unwrapping <code>plyxp</code></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a></a>xp <span class="sc">|&gt;</span> <span class="fu">se</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 3 4 
metadata(0):
assays(1): counts
rownames(3): A B C
rowData names(1): length
colnames(4): a b c d
colData names(1): type</code></pre>
</div>
</div>
</section>
<section id="tidyomics-in-long-read-rna-seq-analysis" class="slide level2">
<h2>tidyomics in long read RNA-seq analysis</h2>
<ul>
<li><p>Filtering, exploration, QC</p></li>
<li><p>Data thinning (count splitting)</p></li>
<li><p>Grouping isoforms for DTU analysis, keeping track of aggregated transcript sets</p></li>
<li><p>Projecting splicing features from transcript-level quantification and analysis to events like skipped exons, retained introns, alternative 5’ / 3’ UTR, etc.</p></li>
</ul>
</section>
<section id="computing-on-genomic-ranges" class="slide level2">
<h2>Computing on genomic ranges</h2>
<ul>
<li><p>Translate biological questions about the genome into code.</p></li>
<li><p>Leverage familiar concepts of filters, joins, grouping, mutation, or summarization.</p></li>
</ul>

<img data-src="figures/fillingthegaps.png" class="quarto-figure quarto-figure-center r-stretch" alt="T2T Consoritium, Science."></section>
<section id="genomic-overlap-as-a-join" class="slide level2">
<h2>Genomic overlap as a join</h2>

<img data-src="figures/woodjoin.png" class="quarto-figure quarto-figure-center r-stretch" alt="Interwoven stacks of wood"></section>
<section id="genomic-overlap-as-a-join-1" class="slide level2">
<h2>Genomic overlap as a join</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a></a><span class="fu">library</span>(plyranges)</span>
<span id="cb53-2"><a></a>x</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 40 ranges and 1 metadata column:
       seqnames    ranges strand |     score
          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
   [1]        1   114-115      * |   4.88780
   [2]        1   129-130      * |   4.40817
   [3]        1   154-155      * |   5.18773
   [4]        1   195-196      * |   5.81901
   [5]        1   200-201      * |   4.14720
   ...      ...       ...    ... .       ...
  [36]        1   898-899      * |   6.55006
  [37]        1   922-923      * |   6.46796
  [38]        1   956-957      * |   4.93079
  [39]        1   957-958      * |   4.34292
  [40]        1   966-967      * |   4.53976
  -------
  seqinfo: 1 sequence from an unspecified genome</code></pre>
</div>
</div>
</section>
<section id="genomic-overlap-as-a-join-2" class="slide level2">
<h2>Genomic overlap as a join</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a></a>y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 1 metadata column:
      seqnames    ranges strand |       id
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
  [1]        1   101-300      * |        a
  [2]        1   451-650      * |        b
  [3]        1  801-1000      * |        c
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="computing-overlap-here-inner-join" class="slide level2">
<h2>Computing overlap (here inner join)</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a></a>x <span class="sc">|&gt;</span> <span class="fu">join_overlap_inner</span>(y)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 30 ranges and 2 metadata columns:
       seqnames    ranges strand |     score       id
          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;factor&gt;
   [1]        1   114-115      * |   4.88780        a
   [2]        1   129-130      * |   4.40817        a
   [3]        1   154-155      * |   5.18773        a
   [4]        1   195-196      * |   5.81901        a
   [5]        1   200-201      * |   4.14720        a
   ...      ...       ...    ... .       ...      ...
  [26]        1   898-899      * |   6.55006        c
  [27]        1   922-923      * |   6.46796        c
  [28]        1   956-957      * |   4.93079        c
  [29]        1   957-958      * |   4.34292        c
  [30]        1   966-967      * |   4.53976        c
  -------
  seqinfo: 1 sequence from an unspecified genome</code></pre>
</div>
</div>
</section>
<section id="chaining-operations" class="slide level2">
<h2>Chaining operations</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a></a>x <span class="sc">|&gt;</span></span>
<span id="cb59-2"><a></a>  <span class="fu">filter</span>(score <span class="sc">&gt;</span> <span class="fl">3.5</span>) <span class="sc">|&gt;</span></span>
<span id="cb59-3"><a></a>  <span class="fu">join_overlap_inner</span>(y) <span class="sc">|&gt;</span></span>
<span id="cb59-4"><a></a>  <span class="fu">group_by</span>(id) <span class="sc">|&gt;</span></span>
<span id="cb59-5"><a></a>  <span class="fu">summarize</span>(<span class="at">ave_score =</span> <span class="fu">mean</span>(score), <span class="at">n =</span> <span class="fu">n</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 3 rows and 3 columns
        id ave_score         n
  &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
1        a   5.00465        10
2        b   5.43353        11
3        c   5.45538         7</code></pre>
</div>
</div>
<p>Options: <code>directed</code>, <code>within</code>, <code>maxgap</code>, <code>minoverlap</code>, etc.</p>
</section>
<section id="pipe-to-plot" class="slide level2">
<h2>Pipe to plot</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a></a>x <span class="sc">|&gt;</span></span>
<span id="cb61-2"><a></a>  <span class="fu">filter</span>(score <span class="sc">&gt;</span> <span class="fl">3.5</span>) <span class="sc">|&gt;</span></span>
<span id="cb61-3"><a></a>  <span class="fu">join_overlap_inner</span>(y) <span class="sc">|&gt;</span></span>
<span id="cb61-4"><a></a>  <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span></span>
<span id="cb61-5"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> id, <span class="at">y =</span> score)) <span class="sc">+</span> </span>
<span id="cb61-6"><a></a>  <span class="fu">geom_violin</span>() <span class="sc">+</span> <span class="fu">geom_jitter</span>(<span class="at">width=</span>.<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="tidy-intro-talk_files/figure-revealjs/ranges-plot-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="convenience-functions" class="slide level2">
<h2>Convenience functions</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a></a>y <span class="sc">|&gt;</span> </span>
<span id="cb62-2"><a></a>  <span class="fu">anchor_5p</span>() <span class="sc">|&gt;</span> <span class="co"># 5', 3', start, end, center</span></span>
<span id="cb62-3"><a></a>  <span class="fu">mutate</span>(<span class="at">width=</span><span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb62-4"><a></a>  <span class="fu">join_nearest</span>(x, <span class="at">distance=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 3 metadata columns:
      seqnames    ranges strand |       id     score  distance
         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
  [1]        1       101      * |        a   4.88780        12
  [2]        1       451      * |        b   3.99047         1
  [3]        1       801      * |        c   6.49877        22
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="manipulating-transcripts-for-long-read-rna-seq" class="slide level2">
<h2>Manipulating transcripts for long read RNA-seq</h2>
<ul>
<li>Bea Campillo in the lab (UNC BCB) was interested in developing a visualization tool after differential transcript usage (DTU)</li>
<li>We run e.g.&nbsp;<a href="https://github.com/COMBINE-lab/oarfish">oarfish</a>, import into R with tximeta, then run <a href="https://www.bioconductor.org/packages/satuRn">satuRn</a> for DTU testing</li>
<li>Bea developed <em>SPLain</em> for visualizing per gene DTU results, also detecting skipped exons and performing sequence analysis using <em>plyranges</em>. Coming soon!</li>
</ul>
</section>
<section id="splain-visualize-splicing-changes" class="slide level2">
<h2>SPLain: visualize splicing changes</h2>

<img data-src="figures/splain1a.png" class="quarto-figure quarto-figure-center r-stretch" alt="SPLain app"></section>
<section id="splain-visualize-splicing-changes-1" class="slide level2">
<h2>SPLain: visualize splicing changes</h2>

<img data-src="figures/splain1b.png" class="quarto-figure quarto-figure-center r-stretch" alt="SPLain app"></section>
<section id="splain-sequence-analysis" class="slide level2">
<h2>SPLain: sequence analysis</h2>

<img data-src="figures/splain2.png" class="quarto-figure quarto-figure-center r-stretch" alt="SPLain app"></section>
<section id="splain-sequence-analysis-1" class="slide level2">
<h2>SPLain: sequence analysis</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a></a>pos_exons  <span class="ot">&lt;-</span> exons_flat <span class="sc">|&gt;</span> <span class="fu">filter</span>(sign <span class="sc">==</span>  <span class="dv">1</span>)</span>
<span id="cb64-2"><a></a>neg_exons <span class="ot">&lt;-</span> exons_flat <span class="sc">|&gt;</span> <span class="fu">filter</span>(sign <span class="sc">==</span>  <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb64-3"><a></a></span>
<span id="cb64-4"><a></a>candidates <span class="ot">&lt;-</span> pos_exons <span class="sc">|&gt;</span></span>
<span id="cb64-5"><a></a>  <span class="fu">filter_by_non_overlaps_directed</span>(neg_exons) <span class="sc">|&gt;</span></span>
<span id="cb64-6"><a></a></span>
<span id="cb64-7"><a></a><span class="co"># ...</span></span>
<span id="cb64-8"><a></a></span>
<span id="cb64-9"><a></a>upstr_downreg_exons <span class="ot">&lt;-</span> downreg_exons <span class="sc">|&gt;</span></span>
<span id="cb64-10"><a></a>  <span class="fu">flank_upstream</span>(<span class="at">width =</span> width_upstream) </span>
<span id="cb64-11"><a></a></span>
<span id="cb64-12"><a></a>seq_downreg_exons <span class="ot">&lt;-</span>  Hsapiens <span class="sc">|&gt;</span></span>
<span id="cb64-13"><a></a>  <span class="fu">getSeq</span>(upstr_downreg_exons) <span class="sc">|&gt;</span></span>
<span id="cb64-14"><a></a>  <span class="fu">RNAStringSet</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="assess-significance-using-a-tidy-grammar" class="slide level2">
<h2>Assess significance using a tidy grammar</h2>
<ul>
<li><p>Asking about the enrichment of variants near genes, or peaks in TADs, often requires a lot of custom R code, lots of loops and control code.</p></li>
<li><p>Hard to read, hard to maintain, hard at submission/publication.</p></li>
<li><p>Instead: use familiar verbs, stacked genomic range sets.</p></li>
</ul>
</section>
<section id="defining-null-genomic-range-sets" class="slide level2">
<h2>Defining null genomic range sets</h2>
<ul>
<li><p>We developed a package, <a href="https://nullranges.github.io/nullranges">nullranges</a>, a modular tool to assist with making genomic comparisons.</p></li>
<li><p>Only provides null genomic range sets for investigating various hypotheses. That’s it!</p></li>
<li><p>Doesn’t do enrichment analysis <em>per se</em>, but can be combined with <em>plyranges</em>, <em>ggplot2</em>, etc.</p></li>
</ul>

<img data-src="figures/nullranges.png" class="quarto-figure quarto-figure-center r-stretch"></section>
<section id="bootstrapping-ranges" class="slide level2">
<h2>Bootstrapping ranges</h2>
<p>Statistical papers from the ENCODE project noted that <em>block bootstrapping</em> genomic data preserves important spatial patterns (Bickel <em>et al.</em> 2010).</p>

<img data-src="figures/boot.png" class="quarto-figure quarto-figure-center r-stretch" alt="Diagram of block bootstrapping"></section>
<section id="bootranges" class="slide level2">
<h2>bootRanges</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a></a><span class="fu">library</span>(nullranges)</span>
<span id="cb65-2"><a></a>boot <span class="ot">&lt;-</span> <span class="fu">bootRanges</span>(x, <span class="at">blockLength=</span><span class="dv">10</span>, <span class="at">R=</span><span class="dv">20</span>)</span>
<span id="cb65-3"><a></a><span class="co"># keep track of bootstrap iteration, gives boot dist'n</span></span>
<span id="cb65-4"><a></a>boot <span class="sc">|&gt;</span></span>
<span id="cb65-5"><a></a>  <span class="fu">join_overlap_inner</span>(y) <span class="sc">|&gt;</span></span>
<span id="cb65-6"><a></a>  <span class="fu">group_by</span>(iter, id) <span class="sc">|&gt;</span> <span class="co"># bootstrap iter, range id</span></span>
<span id="cb65-7"><a></a>  <span class="fu">summarize</span>(<span class="at">n_overlaps =</span> <span class="fu">n</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 60 rows and 3 columns
     iter       id n_overlaps
    &lt;Rle&gt; &lt;factor&gt;  &lt;integer&gt;
1       1        a          4
2       1        b         10
3       1        c          4
4       2        a         11
5       2        b         10
...   ...      ...        ...
56     19        b         18
57     19        c         11
58     20        a         10
59     20        b         15
60     20        c          8</code></pre>
</div>
</div>
</section>
<section id="bootranges-1" class="slide level2">
<h2>bootRanges</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a></a>boot <span class="sc">|&gt;</span></span>
<span id="cb67-2"><a></a>  <span class="fu">join_overlap_inner</span>(y) <span class="sc">|&gt;</span></span>
<span id="cb67-3"><a></a>  <span class="fu">group_by</span>(iter, id) <span class="sc">|&gt;</span> <span class="co"># bootstrap iter, range id</span></span>
<span id="cb67-4"><a></a>  <span class="fu">summarize</span>(<span class="at">n_overlaps =</span> <span class="fu">n</span>()) <span class="sc">|&gt;</span></span>
<span id="cb67-5"><a></a>  <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span></span>
<span id="cb67-6"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> id, <span class="at">y =</span> n_overlaps)) <span class="sc">+</span> </span>
<span id="cb67-7"><a></a>  <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="tidy-intro-talk_files/figure-revealjs/boot-plot-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="matching-ranges" class="slide level2">
<h2>Matching ranges</h2>
<p>Matching on covariates from a large pool allows for more focused hypothesis testing.</p>

<img data-src="figures/match.png" class="quarto-figure quarto-figure-center r-stretch" alt="Diagram of matching genomic ranges"></section>
<section id="matchranges" class="slide level2">
<h2>matchRanges</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a></a>xprime <span class="ot">&lt;-</span> x <span class="sc">|&gt;</span></span>
<span id="cb68-2"><a></a>  <span class="fu">filter</span>(score <span class="sc">&gt;</span> <span class="dv">5</span>) <span class="sc">|&gt;</span> <span class="co"># xprime are hi-score</span></span>
<span id="cb68-3"><a></a>  <span class="fu">mutate</span>(<span class="at">score =</span> <span class="fu">rnorm</span>(<span class="fu">n</span>(), <span class="at">mean =</span> score, <span class="at">sd =</span> .<span class="dv">5</span>)) <span class="co"># jitter data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="matchranges-1" class="slide level2">
<h2>matchRanges</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a></a>m <span class="ot">&lt;-</span> <span class="fu">matchRanges</span>(<span class="at">focal =</span> xprime, <span class="at">pool =</span> x, <span class="at">covar =</span> <span class="sc">~</span>score)</span>
<span id="cb69-2"><a></a><span class="fu">plotCovariate</span>(m)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="tidy-intro-talk_files/figure-revealjs/match-covar-plot-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="bind-ranges" class="slide level2">
<h2>Bind ranges</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a></a>combined <span class="ot">&lt;-</span> <span class="fu">bind_ranges</span>(</span>
<span id="cb70-2"><a></a>  <span class="at">focal =</span> xprime,</span>
<span id="cb70-3"><a></a>  <span class="at">matched =</span> <span class="fu">matched</span>(m),</span>
<span id="cb70-4"><a></a>  <span class="at">pool =</span> x,</span>
<span id="cb70-5"><a></a>  <span class="at">.id =</span> <span class="st">"origin"</span></span>
<span id="cb70-6"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is now “tidy data” with the two group concatenated and a new metadata column, <code>origin</code>.</p>
</section>
<section id="group-and-summarize" class="slide level2">
<h2>Group and summarize</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a></a>combined <span class="sc">|&gt;</span> </span>
<span id="cb71-2"><a></a>  <span class="fu">join_overlap_inner</span>(y) <span class="sc">|&gt;</span></span>
<span id="cb71-3"><a></a>  <span class="fu">group_by</span>(id, origin) <span class="sc">|&gt;</span></span>
<span id="cb71-4"><a></a>  <span class="fu">summarize</span>(<span class="at">ave_score =</span> <span class="fu">mean</span>(score))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 9 rows and 3 columns
        id  origin ave_score
  &lt;factor&gt;   &lt;Rle&gt; &lt;numeric&gt;
1        a   focal   6.17495
2        a matched   5.54412
3        a    pool   5.00465
4        b   focal   5.79665
5        b matched   6.07625
6        b    pool   5.25055
7        c   focal   6.32536
8        c matched   5.82463
9        c    pool   5.14840</code></pre>
</div>
</div>
</section>
<section id="pipe-to-plot-1" class="slide level2">
<h2>Pipe to plot</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a></a>combined <span class="sc">|&gt;</span> </span>
<span id="cb73-2"><a></a>  <span class="fu">join_overlap_inner</span>(y) <span class="sc">|&gt;</span></span>
<span id="cb73-3"><a></a>  <span class="fu">group_by</span>(id, origin) <span class="sc">|&gt;</span></span>
<span id="cb73-4"><a></a>  <span class="fu">summarize</span>(<span class="at">ave_score =</span> <span class="fu">mean</span>(score), <span class="at">sd =</span> <span class="fu">sd</span>(score)) <span class="sc">|&gt;</span></span>
<span id="cb73-5"><a></a>  <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span></span>
<span id="cb73-6"><a></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(origin, ave_score, </span>
<span id="cb73-7"><a></a>             <span class="at">ymin=</span>ave_score<span class="dv">-2</span><span class="sc">*</span>sd, <span class="at">ymax=</span>ave_score<span class="sc">+</span><span class="dv">2</span><span class="sc">*</span>sd)) <span class="sc">+</span> </span>
<span id="cb73-8"><a></a>  <span class="fu">geom_pointrange</span>() <span class="sc">+</span></span>
<span id="cb73-9"><a></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>id, <span class="at">labeller =</span> label_both)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="tidy-intro-talk_files/figure-revealjs/match-overlap-plot-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section id="bootranges-and-matchranges-methods" class="slide level2">
<h2>bootRanges and matchRanges methods</h2>
<p>Published as Application Notes in 2023:</p>
<ul>
<li><p><a href="https://doi.org/10.1093/bioinformatics/btad190">bootRanges</a></p></li>
<li><p><a href="https://doi.org/10.1093/bioinformatics/btad197">matchRanges</a></p></li>
</ul>
</section>
<section id="current-state-and-future-directions" class="slide level2">
<h2>Current state and future directions</h2>
<ul>
<li><p>Writing up <code>plyxp</code> as an Application Note.</p></li>
<li><p>Have shown it locally and to an industry team.</p></li>
<li><p>Working with Stefano Mangiola’s team on consistent printing, messaging.</p></li>
<li><p>Working on tutorials, workshops, documentation.</p></li>
</ul>
</section>
<section id="what-has-been-implemented" class="slide level2">
<h2>What has been implemented</h2>
<ul>
<li><p>Matrix-shaped objects (SE, SCE) - Mangiola <em>et al.</em></p></li>
<li><p>Ranges - Lee <em>et al.</em></p></li>
<li><p>Interactions - Serizay <em>et al.</em></p></li>
<li><p>Cytometry - Keyes <em>et al.</em></p></li>
<li><p>Spatial - Hutchison <em>et al.</em></p></li>
<li><p>more to come…</p></li>
</ul>
</section>
<section id="ongoing-work" class="slide level2">
<h2>Ongoing work</h2>
<ul>
<li>Package code and non-standard evaluation (IMO I don’t recommend)</li>
<li>Optimized code</li>
<li>Conflicts</li>
<li>Documentation!</li>
</ul>
</section>
<section id="how-to-contribute" class="slide level2">
<h2>How to contribute</h2>
<ul>
<li><p><a href="https://github.com/tidyomics" class="uri">https://github.com/tidyomics</a> open challenges</p></li>
<li><p>For project motivation, read the <a href="https://www.biorxiv.org/content/10.1101/2023.09.10.557072v2">paper</a></p></li>
<li><p><code>#tidiness_in_bioc</code> channel in Bioconductor Slack</p></li>
<li><p>More complicated use cases: <a href="https://tidyomics.github.io/tidy-ranges-tutorial">Tidy ranges tutorial</a></p></li>
</ul>
</section>
<section id="omics-data-analysis-requires-looking-at" class="slide level2">
<h2>Omics data analysis requires looking at:</h2>
<ul>
<li>main contributions to variance (e.g.&nbsp;PCA, see <code>plotPCA</code> for bulk and <a href="https://bioconductor.org/books/release/OSCA/">OSCA</a> for sc), also <code>variancePartition</code></li>
<li>column and row densities (<code>geom_density</code> of select rows/columns, or <code>geom_violin</code>)</li>
<li>known positive features, feature-level plots (<code>filter</code> to feature, pipe to <code>geom_point</code> etc.)</li>
</ul>
</section>
<section id="acknowledgments" class="slide level2">
<h2>Acknowledgments</h2>
<style>
  .small-text p {
    font-size: 0.6em;
  }
</style>
<ul>
<li><p>Stefano Mangiola (co-lead, tidySE, tidySCE, spatial)</p></li>
<li><p>Justin Landis (<em>plyxp</em>) and Beatriz Campillo Minano (<em>SPLain</em>)</p></li>
<li><p>Eric Davis, Wancen Mu, Doug Phanstiel (<em>nullranges</em>)</p></li>
<li><p>Stuart Lee, M. Lawrence, D. Cook (<em>plyranges</em> developers)</p></li>
</ul>
<div class="small-text">
<p><strong>tidyomics developers</strong>: William Hutchison, Timothy Keyes, Helena Crowell, Jacques Serizay, Charlotte Soneson, Eric Davis, Noriaki Sato, Lambda Moses, Boyd Tarlinton, Abdullah Nahid, Miha Kosmac, Quentin Clayssen, Victor Yuan, Wancen Mu, Ji-Eun Park, Izabela Mamede, Min Hyung Ryu, Pierre-Paul Axisa, Paulina Paiz, Chi-Lam Poon, Ming Tang</p>
<p><em>tidyomics</em> project funded by an Essential Open Source Software award from CZI and Wellcome Trust</p>
</div>


</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="site_libs/revealjs/plugin/search/search.js"></script>
  <script src="site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>